### PR TITLE
Avoid managed space leak

### LINKED ISSUE #4: Investigate whether using Managed could be an issue
Judging by this note from the [documentation](https://hackage.haskell.org/package/managed-1.0.7/docs/Control-Monad-Managed.html).

> NOTE: Managed may leak space if used in an infinite loop like this example.

We might run into this issue in the long-term.

### PR DESCRIPTION
**Pull request by gvolpe**
closes #4 

Big refactor to avoid running long-lived computations in the `Managed` monad, to avoid space leaks.

- The `Pulsar` monad now wraps a `ReaderT PulsarCtx IO a`.
- A new `Connection` monad that wraps `Managed a` was introduced.
- Removed the `unliftio` dependency.

They are both combined when calling `runPulsar` or `runPulsar'`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -15,17 +15,19 @@ The example located in `test/Main.hs` showcases a consumer & producer running co
...
+  liftIO $ program c p
+
+conn :: PulsarConnection
+conn = connect defaultConnectData
+
+program :: Consumer IO -> Producer IO -> IO ()
+program (Consumer fetch ack) (Producer send) =
   let c = forever $ fetch >>= \(Message i m) -> msgDecoder m >> ack i
-      p = forever $ sleep 5 >> traverse_ produce messages
+      p = forever $ sleep 5 >> traverse_ send messages
```
Comment (psisoyev) [inline README.md:29]: why this sleep is needed?

Comment (gvolpe) [reply README.md:29]: @psisoyev you need to sleep FFS :smile:

Comment (gvolpe) [reply README.md:29]: That `sleep` is just to generate messages every 5 seconds, forever. It makes for a better example IMO but it is not needed.

Comment (psisoyev) [reply README.md:29]: shouldn't we send out the message as soon as we have it?

Comment (gvolpe) [reply README.md:29]: We could but we need to guarantee we are consuming before start sending events and that cannot be guaranteed by using `concurrently` this way, it is a very naive program...

Comment (gvolpe) [reply README.md:29]: There are still a few rough edges I need to work through to make it more robust, this PR is a step towards that goal

**Review by psisoyev — APPROVED** [review]

**Pull request merged**